v3/lint/configuration.go

```diff
@@ -0,0 +1,242 @@
+	return c.deserializeConfigInto(lint, namespace)
+}
+
+// NewConfig attempts to instantiate a configuration by consuming the contents of the provided reader.
+//
+// The contents of the provided reader MUST be in a valid TOML format. The caller of this function
+// is responsible for closing the reader, if appropriate.
+func NewConfig(r io.Reader) (Configuration, error) {
+	tree, err := toml.LoadReader(r)
+	if err != nil {
+		return Configuration{}, err
+	}
+	return Configuration{tree}, nil
+}
+
+// NewConfigFromFile attempts to instantiate a configuration from the provided filesystem path.
+//
+// The file pointed to by `path` MUST be valid TOML file. If `path` is the empty string then
+// an empty configuration is returned.
+func NewConfigFromFile(path string) (Configuration, error) {
+	if path == "" {
+		return NewEmptyConfig(), nil
+	}
+	f, err := os.Open(path)
+	if err != nil {
+		return Configuration{}, fmt.Errorf("failed to open the provided configuration at %s. Error: %s", path, err.Error())
+	}
+	defer f.Close()
+	return NewConfig(f)
+}
+
+// NewConfigFromString attempts to instantiate a configuration from the provided string.
+//
+// The provided string MUST be in a valid TOML format.
+func NewConfigFromString(config string) (Configuration, error) {
+	return NewConfig(strings.NewReader(config))
+}
+
+// NewEmptyConfig returns a configuration that is backed by an entirely empty TOML tree.
+//
+// This is useful if no particular configuration is set at all by the user of ZLint as
+// any attempt to resolve a namespace in `deserializeConfigInto` fails and thus results
+// in all defaults for all lints being maintained.
+func NewEmptyConfig() Configuration {
+	ctx, _ := NewConfigFromString("")
+	return ctx
+}
+
+// deserializeConfigInto deserializes the section labeled by the provided `namespace`
+// into the provided target `interface{}`.
+//
+// For example, given the following configuration...
+//
+// ```
+// [e_some_lint]
+// field = 1
+// flag = false
+//
+// [w_some_other_lint]
+// is_web_pki = true
+// ```
+//
+// And the following struct definition...
+//
+// ```
+// type SomeOtherLint {
+//		IsWebPKI bool `toml:"is_web_pki"`
+// }
+// ```
+//
+// Then the invocation of this function should be...
+//
+// ```
+// lint := &SomeOtherLink{}
+// deserializeConfigInto(link, "w_some_other_lint")
+// ```
+//
+// If there is no such namespace found in this configuration then provided the namespace specific data encoded
+// within `target` is left unmodified. However, configuration of higher scoped fields will still be attempted.
+func (c Configuration) deserializeConfigInto(target interface{}, namespace string) error {
+	if tree := c.tree.Get(namespace); tree != nil {
+		err := tree.(*toml.Tree).Unmarshal(target)
+		if err != nil {
+			return err
+		}
+	}
+	return c.resolveHigherScopedReferences(target)
+}
+
+// resolveHigherScopeReferences takes in an interface{} value and attempts to
```

Comment:
Indeed, this method is what accomplished the somewhat fancy scoping feature by reflecting into the struct provided by the lint and trying to find and data member that is one of the higher scoped configuration fields.